File: backend/src/models/promotion.model.js

const mongoose = require('mongoose');

/**
 * @swagger
 * components:
 *   schemas:
 *     Promotion:
 *       type: object
 *       required:
 *         - title
 *         - description
 *         - discountPercentage
 *         - startDate
 *         - endDate
 *         - imageUrl
 *         - residenceId
 *       properties:
 *         id:
 *           type: string
 *           description: L'ID auto-généré de la promotion
 *         title:
 *           type: string
 *           description: Le titre de la promotion
 *         description:
 *           type: string
 *           description: La description détaillée de la promotion
 *         discountPercentage:
 *           type: number
 *           description: Le pourcentage de réduction offert
 *         discountCode:
 *           type: string
 *           description: Le code promo à utiliser (optionnel)
 *         startDate:
 *           type: string
 *           format: date
 *           description: La date de début de la promotion
 *         endDate:
 *           type: string
 *           format: date
 *           description: La date de fin de la promotion
 *         imageUrl:
 *           type: string
 *           description: L'URL de l'image illustrant la promotion
 *         residenceId:
 *           type: string
 *           description: L'ID de la résidence concernée par la promotion
 *         badge:
 *           type: string
 *           description: Un texte court pour le badge de la promotion (optionnel)
 *         isExclusive:
 *           type: boolean
 *           description: Indique si la promotion est exclusive
 *         type:
 *           type: string
 *           enum: [discount, flash, seasonal, bundle, exclusive, newUser]
 *           description: Le type de promotion
 *         termsAndConditions:
 *           type: string
 *           description: Conditions d'utilisation de la promotion (optionnel)
 *       example:
 *         id: 61dbae02-c147-4e28-863c-db7bd402b2d6
 *         title: Week-end luxueux à -30%
 *         description: Profitez d'un séjour de luxe avec une réduction exclusive de 30% sur nos suites premium.
 *         discountPercentage: 30
 *         discountCode: WEEKEND30
 *         startDate: 2025-04-01T00:00:00.000Z
 *         endDate: 2025-04-30T23:59:59.000Z
 *         imageUrl: https://example.com/images/promotion1.jpg
 *         residenceId: 5f8a7d6c5e8d3a2b1c9f8a7d
 *         badge: PROMO FLASH
 *         isExclusive: true
 *         type: flash
 *         termsAndConditions: Offre valable pour les réservations de plus de 2 nuits, non cumulable avec d'autres promotions.
 */

const promotionSchema = mongoose.Schema({
  title: {
    type: String,
    required: [true, 'Le titre est requis'],
    trim: true,
    maxlength: [100, 'Le titre ne peut pas dépasser 100 caractères']
  },
  description: {
    type: String,
    required: [true, 'La description est requise'],
    trim: true
  },
  discountPercentage: {
    type: Number,
    required: [true, 'Le pourcentage de réduction est requis'],
    min: [0, 'Le pourcentage ne peut pas être négatif'],
    max: [100, 'Le pourcentage ne peut pas dépasser 100%']
  },
  discountCode: {
    type: String,
    trim: true
  },
  startDate: {
    type: Date,
    required: [true, 'La date de début est requise']
  },
  endDate: {
    type: Date,
    required: [true, 'La date de fin est requise']
  },
  imageUrl: {
    type: String,
    required: [true, 'L\'URL de l\'image est requise']
  },
  residenceId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Residence',
    required: [true, 'L\'ID de la résidence est requis']
  },
  badge: {
    type: String,
    trim: true
  },
  isExclusive: {
    type: Boolean,
    default: false
  },
  type: {
    type: String,
    enum: ['discount', 'flash', 'seasonal', 'bundle', 'exclusive', 'newUser'],
    default: 'discount'
  },
  termsAndConditions: {
    type: String
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  },
  createdBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  updatedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Indexation pour améliorer les performances de recherche
promotionSchema.index({ title: 'text', description: 'text' });
promotionSchema.index({ residenceId: 1 });
promotionSchema.index({ type: 1 });
promotionSchema.index({ isExclusive: 1 });
promotionSchema.index({ startDate: 1, endDate: 1 });

// Vérifier si la promotion est active à une date donnée
promotionSchema.methods.checkActive = function(date = new Date()) {
  return date >= this.startDate && date <= this.endDate;
};

// Calculer le prix après réduction
promotionSchema.methods.calculateDiscountedPrice = function(originalPrice) {
  return originalPrice * (1 - this.discountPercentage / 100);
};

// Calculer le temps restant avant expiration (retourne un objet avec jours, heures, minutes)
promotionSchema.methods.getTimeRemaining = function() {
  const now = new Date();
  const difference = this.endDate - now;
  
  // Si la promotion est déjà expirée
  if (difference <= 0) {
    return { days: 0, hours: 0, minutes: 0, expired: true };
  }
  
  const days = Math.floor(difference / (1000 * 60 * 60 * 24));
  const hours = Math.floor((difference % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
  const minutes = Math.floor((difference % (1000 * 60 * 60)) / (1000 * 60));
  
  return { days, hours, minutes, expired: false };
};

// Virtuel : indique si c'est une offre de dernière minute (moins de 24h restantes)
promotionSchema.virtual('isLastMinute').get(function() {
  const remaining = this.getTimeRemaining();
  return !remaining.expired && remaining.days === 0 && remaining.hours < 24;
});

// Middleware pre-save pour s'assurer que endDate est après startDate
promotionSchema.pre('save', function(next) {
  if (this.endDate <= this.startDate) {
    return next(new Error('La date de fin doit être postérieure à la date de début'));
  }
  next();
});

const Promotion = mongoose.model('Promotion', promotionSchema);

module.exports = Promotion;

File: chapechape_partner/lib/core/models/promotion/promotion_model.dart
import 'package:equatable/equatable.dart';
import '../residence/residence.dart';

/// Types de promotions disponibles
enum PromotionType {
  discount, // Réduction simple sur le prix
  flash,    // Promotion de courte durée
  seasonal, // Promotion saisonnière
  bundle,   // Bundle d'offres
  exclusive, // Offre exclusive
  newUser   // Promotion pour nouveaux utilisateurs
}

/// Extension pour convertir le type de promotion en string
extension PromotionTypeExtension on PromotionType {
  String get value {
    switch (this) {
      case PromotionType.discount: return 'discount';
      case PromotionType.flash: return 'flash';
      case PromotionType.seasonal: return 'seasonal';
      case PromotionType.bundle: return 'bundle';
      case PromotionType.exclusive: return 'exclusive';
      case PromotionType.newUser: return 'newUser';
    }
  }
  
  static PromotionType fromString(String value) {
    switch (value) {
      case 'discount': return PromotionType.discount;
      case 'flash': return PromotionType.flash;
      case 'seasonal': return PromotionType.seasonal;
      case 'bundle': return PromotionType.bundle;
      case 'exclusive': return PromotionType.exclusive;
      case 'newUser': return PromotionType.newUser;
      default: return PromotionType.discount;
    }
  }
}

/// Modèle représentant une promotion ou offre spéciale
class PromotionModel extends Equatable {
  final String id;
  final String title;
  final String description;
  final PromotionType type;
  final double discountPercentage;
  final double? discountAmount;
  final DateTime startDate;
  final DateTime endDate;
  final bool isActive;
  final bool isExclusive;
  final String? residenceId;
  final Residence? residence;
  final String? code;
  final int usageLimit;
  final int usageCount;
  final String? imageUrl;
  
  const PromotionModel({
    required this.id,
    required this.title,
    required this.description,
    required this.type,
    required this.discountPercentage,
    this.discountAmount,
    required this.startDate,
    required this.endDate,
    required this.isActive,
    required this.isExclusive,
    this.residenceId,
    this.residence,
    this.code,
    required this.usageLimit,
    required this.usageCount,
    this.imageUrl,
  });
  
  @override
  List<Object?> get props => [
    id, title, description, type, discountPercentage, discountAmount,
    startDate, endDate, isActive, isExclusive, residenceId, residence,
    code, usageLimit, usageCount, imageUrl
  ];
  
  /// Crée un objet PromotionModel à partir d'une réponse JSON de l'API
  factory PromotionModel.fromJson(Map<String, dynamic> json) {
    return PromotionModel(
      id: json['_id'] ?? json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      type: json['type'] != null 
          ? PromotionTypeExtension.fromString(json['type']) 
          : PromotionType.discount,
      discountPercentage: json['discountPercentage']?.toDouble() ?? 0.0,
      discountAmount: json['discountAmount']?.toDouble(),
      startDate: json['startDate'] != null 
          ? DateTime.parse(json['startDate']) 
          : DateTime.now(),
      endDate: json['endDate'] != null 
          ? DateTime.parse(json['endDate']) 
          : DateTime.now().add(const Duration(days: 30)),
      isActive: json['isActive'] ?? false,
      isExclusive: json['isExclusive'] ?? false,
      residenceId: json['residenceId'],
      residence: json['residence'] != null 
          ? Residence.fromJson(json['residence']) 
          : null,
      code: json['code'],
      usageLimit: json['usageLimit'] ?? 0,
      usageCount: json['usageCount'] ?? 0,
      imageUrl: json['imageUrl'],
    );
  }
  
  /// Convertit l'objet en format JSON pour l'API
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'type': type.value,
      'discountPercentage': discountPercentage,
      if (discountAmount != null) 'discountAmount': discountAmount,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate.toIso8601String(),
      'isActive': isActive,
      'isExclusive': isExclusive,
      if (residenceId != null) 'residenceId': residenceId,
      if (code != null) 'code': code,
      'usageLimit': usageLimit,
      'usageCount': usageCount,
      if (imageUrl != null) 'imageUrl': imageUrl,
    };
  }
  
  /// Crée une copie de l'objet avec certains champs modifiés
  PromotionModel copyWith({
    String? id,
    String? title,
    String? description,
    PromotionType? type,
    double? discountPercentage,
    double? discountAmount,
    DateTime? startDate,
    DateTime? endDate,
    bool? isActive,
    bool? isExclusive,
    String? residenceId,
    Residence? residence,
    String? code,
    int? usageLimit,
    int? usageCount,
    String? imageUrl,
  }) {
    return PromotionModel(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      type: type ?? this.type,
      discountPercentage: discountPercentage ?? this.discountPercentage,
      discountAmount: discountAmount ?? this.discountAmount,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      isActive: isActive ?? this.isActive,
      isExclusive: isExclusive ?? this.isExclusive,
      residenceId: residenceId ?? this.residenceId,
      residence: residence ?? this.residence,
      code: code ?? this.code,
      usageLimit: usageLimit ?? this.usageLimit,
      usageCount: usageCount ?? this.usageCount,
      imageUrl: imageUrl ?? this.imageUrl,
    );
  }
  
  /// Vérifie si la promotion est en cours (entre les dates de début et de fin)
  bool get isCurrentlyActive {
    final now = DateTime.now();
    return isActive && now.isAfter(startDate) && now.isBefore(endDate);
  }
  
  /// Calcule le montant de la réduction pour un prix donné
  double calculateDiscount(double price) {
    if (discountAmount != null) {
      return discountAmount!;
    }
    return price * (discountPercentage / 100);
  }
  
  /// Calcule le prix après réduction
  double calculateDiscountedPrice(double originalPrice) {
    final discount = calculateDiscount(originalPrice);
    return originalPrice - discount;
  }
}
